### ISSUE TITLE
Importing Vega datasets

### ISSUE THREAD

**Issue by ijlyttle** [ijlyttle/altair#25]
As discussed in #24, it would be useful to have ready access to the Vega datasets. As @AliciaSchep shows, reticulate opens a door for us:

```{r}
data <- reticulate::import("vega_datasets")$data
```

Another approach is to use that door to build a separate package, a path @haleyjeppson has started down for her own use: [haleyjeppson/vegadatasets](https://github.com/haleyjeppson/vegadatasets)

Although I like the idea of a package, the problem I see (at least as far as CRAN is concerned) will be to [document](http://r-pkgs.had.co.nz/data.html#documenting-data) those datasets.

I wonder if we can use reticulate as a loophole. If we did something like this in [zzz.R](https://github.com/ijlyttle/altair/blob/master/R/zzz.R):

```r
alt <- NULL
vegadata <- NULL

.onLoad <- function(libname, pkgname) {
  alt <<- reticulate::import("altair", delay_load = TRUE)
  vegadata <<- reticulate::import("vega_datasets", delay_load = TRUE)$data
}
```

Then, "legally" (I think), we would just have to document the `vegadata` object (or whatever we call it).

What do you all think? If you like it, any name-suggestions beyond `vegadata`?  Maybe `vega_data`?

**Comment by AliciaSchep** [comment]
I think the reticulate based method would be the easiest method, and the approach you document above seems like a good one! My understanding is that when you install Altair it also installs vega_datasets (or at least I never separately installed it...) so it seems reasonable to use reticulate in this way within this package.  `vega_data` seems like a good name.

I think the R package would be useful if the intention was to make these data sets accessible to the broader R community and not just altair users.  These data sets might be useful for other visualization packages, and then someone wouldn't need to get python & Altair installed to use them.

**Comment by ijlyttle** [comment]
I have added `vega_data` into the master branch. It seems to work OK for me, except for the remote datasets, where I get this error:

```
 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:777)>
```

I strongly suspect it is because I have not set up the proxy for my Python installation. As well, it looks like @haleyjeppson succeeded to get the remote data in [haleyjeppson/vegadatasets](https://github.com/haleyjeppson/vegadatasets).

Here's my example code:

```R
# To list available datasets
vega_data$list_datasets()

# When accessing a dataset, substitute any "-" in the name with a "_"
head(vega_data$sf_temps())

# Metadata are available for each dataset
vega_data$sf_temps$description
vega_data$sf_temps$url

# For local datasets, local path is available
vega_data$sf_temps$filepath
``` 

If anyone gets a chance, could I impose on you to try this out to make sure it works OK for you?

As well, could you try `vega_data$gapminder()`, just to make sure that it's me?

All being well, I will improve the documentation, then close this issue.

**Comment by ijlyttle** [comment]
Half a solution:

This might be a Mac issue with certificates. My remote-data problem seems to be a certificate problem that gets solved if I set these environment variables in my `.bash_profile`.

```
export SSL_CERT_FILE=/usr/local/etc/openssl/cert.pem
export REQUESTS_CA_BUNDLE=/usr/local/etc/openssl/cert.pem
```

From Python, I can do this:

```python
from vega_datasets import data

data.gapminder()
```

However, I still have the same problem using reticulate.

**Comment by AliciaSchep** [comment]
Not just a Mac issue -- tried it with Ubuntu and got same error with gapminder.

**Comment by ijlyttle** [comment]
Thanks for checking!

Some good news and some bad news:

Bad news:

-  Doing this in `.onload()` is bad

    ```R
    vega_data <<- reticulate::import("vega_datasets", delay_load = TRUE)$data
    ```

    By using `$data`, it makes a reference to the `vega_data` object, making `delay_load = TRUE` moot. I am thinking to ask the reticulate folks if there is a way to handle this `from foo import bar` while maintaining the delay.

Good news: 

- I have found a way to access non-local data. It involves telling Python where to find your certificate and how to use your proxy. This is now in an article I am working on, "Field Guide to Python issues", in my `new-docs` branch, which is going more-slowly than I would like (hey, but I'm learning a lot ¯\_(ツ)_/¯).

More bad news:

- Now, my issue with `vega_data$gapminder()` is different:

    ```
    Error in py_get_attr_impl(x, name, silent) : 
      AttributeError: 'function' object has no attribute 'as_matrix'
    ```

    This may be another question for the reticulate folks.

**Comment by ijlyttle** [comment]
Some updates:

I have chased down the "gapminder" issue; I think that reticulate has a problem if a Pandas DataFrame has a column called `pop` (https://github.com/rstudio/reticulate/issues/244)

I am also trying to see if something can be done in the vega_datasets Python package, so that this would work in `.onLoad()` (https://github.com/altair-viz/vega_datasets/issues/9)

```r
vega_data <<- reticulate::import("vega_datasets.data", delay_load = TRUE)
```

**Comment by ijlyttle** [comment]
@AliciaSchep Thanks again for seeing and sorting out the reticulate issue on reserved-words in DataFrames!

For this issue, I think we have a couple of choices (you may see more, so please pipe in):

1. In `.onload()`, do something like:

    ```r
    vega_data <<- reticulate::import("vega_datasets", delay_load = TRUE)
    ```

    The consequence being that we would refer to `vega_data$data`, whereas before we would refer to `vega_data`.

2. Raise an issue with reticulate to see if there is a way to support importing an object from a module while maintaining `delay_load`.

I know that option (1) will work, but is a bit verbose. I am happy to follow option (2), then use option (1) if it is not feasible.

**Comment by AliciaSchep** [comment]
I think either of those two are good options!  One alternative -- have a `import_vega_datasets` function that does `vega_data <<- reticulate::import("vega_datasets", delay_load = TRUE)$data` rather than it being in `.onload()` (or maybe just `<-` and return).  The user would then have to call that explicitly to get access to the data sets.  That or something else that is a bit more explicit might be appropriate for this use case in that these examples are mostly useful for tutorials and documentation.

It also looks like they are considering add the datasets into Altair: https://github.com/altair-viz/altair/issues/796. That would make this easier!

**Comment by ijlyttle** [comment]
I like the idea of the `import_vega_datasets()` function!

It gets us out of our current predicament, preserving our "current" `vega_data` status.

If, in time, reticulate can offer option (2) - then we can deprecate the function. Same thing if altair-viz/altair#796 comes to pass.

I will implement forthwith!

**Comment by ijlyttle** [comment]
I am running into all sorts of trouble getting 

```r
vega_data <<- reticulate::import("vega_datasets")$data
```

outside of `.onLoad()`. It seems that once the namespace is loaded, then it is sealed. I tried to do some unnatural things like `unlockBinding()` and `lockBinding()`, but could not get it to work.

In the short term, in the absence of a clever solution, I propose not to export `vega_data`, but rather have the user create it explicity:

```r
vega_data <- import_vega_data()
```

and we provide:

```r
import_vega_data <- function() {
  reticulate::import("vega_datasets")$data
}
```

I think the workflow remains the same if/when the function gets deprecated.

**Comment by AliciaSchep** [comment]
Yeah hadn't really thought through whether the `<<-` would be a problem in this context... I think the function you've proposed is a good option!

**Comment by ijlyttle** [comment]
Thanks!

Just to be perfectly clear, my "understanding" came about only in the course of trying to get this to work... 😬

**Comment by ijlyttle** [comment]
merged in #37
